### ISSUE TITLE
Image path replaces

### ISSUE THREAD

**Issue by pekka2** [opencart/opencart#7080]
admin/model/tool/image.php line 4:

`if (!is_file(DIR_IMAGE . $filename) || substr(str_replace('\\', '/', realpath(DIR_IMAGE . $filename)), 0, strlen(DIR_IMAGE)) != str_replace('\\', '/', DIR_IMAGE)) {`

These replace scripts 

`str_replace('\\', '/', realpath(DIR_IMAGE . $filename))`

`str_replace('\\', '/', DIR_IMAGE)`

are completely unnecessary(?)

Path `define('DIR_IMAGE', 'C:\MAMP\htdocs\3020/image/') ` work successfully. But current replace scripts at files image.php and filemanager.php is not work all servers.

**Comment by MrPlow254** [comment]
In Linux, the path separator is /. In Windows, it is either \ or /. So to be universal it needs to use forward slashes.

**Comment by pekka2** [comment]
There is no question in this chain, what is universal script, but works with Opencart scripts.
Latest Openacrt version, where image path scripts work all servers, was Opencart 2.2

**Comment by MrPlow254** [comment]
@pekka2 Can you rephrase that please? I am not following what the issue is.

**Comment by ADDCreative** [comment]
The path checks have probably been added for a good reason. It's probably best not to remove them.

Your issue could be caused by some sort of symbolic link, which cause realpath a different path to your config. See https://github.com/copona/copona/issues/128

You could try something like.
```php
if (!is_file(DIR_IMAGE . $filename) || str_replace('\\', '/', realpath(DIR_IMAGE) . '/' . $filename) !== str_replace('\\'. '/', realpath(DIR_IMAGE . $filename))) {
```

**Comment by pekka2** [comment]
> It's probably best not to remove them.

I am removed it few server, where it has not worked :)

**Comment by pekka2** [comment]
DIR_IMAGE is Ok, realpath is not tested.

**Comment by ADDCreative** [comment]
Are you able to test ```realpath(DIR_IMAGE);```?

**Comment by pekka2** [comment]
Path is sulfur, path begins at document root. Missing diectories '/sites/domain.nl/'

**Comment by osworx** [comment]
@pekka2 
1. it is sometimes very hard to understand your english
2. it would be more useful if you would write more about the used (and failed) environement (which server, which php-version, etc.)
3, if explicite asked, a straight answer would help more than answering scriptic .. (e.g. what is DIR_IMAGE set)
thx!

**Comment by ADDCreative** [comment]
@pekka2 So you are saying ```DIR_IMAGE = '/sites/domain.nl/sulfur/image'``` but ```realpath(DIR_IMAGE) = '/sulfur/image'```?

**Comment by pekka2** [comment]
@ADDCreative 
>So you are saying `DIR_IMAGE = '/sites/domain.nl/sulfur/image'` but `realpath(DIR_IMAGE) = '/sulfur/image'`?

Or `echo DIR_IMAGE;` is '/sites/domain.nl/www/image/' :)

**Comment by danielkerr** [comment]
I should do the str_replace at the config level.

**Comment by ADDCreative** [comment]
It seems to be that ```realpath(DIR_IMAGE)``` returns a different path to ```DIR_IMAGE``` on its own. Possibly due to symbolic links being used on the server.

So DIR_IMAGE is set to something like '/sites/domain.nl/www/image/', but realpath(DIR_IMAGE) returns something like '/sulfur/image'.

@pekka2 Try replacing admin/model/tool/image.php line 4 with the following.
```php
if (!is_file(DIR_IMAGE . $filename) || str_replace('\\', '/', realpath(DIR_IMAGE) . '/' . $filename) !== str_replace('\\'. '/', realpath(DIR_IMAGE . $filename))) {
```php

**Comment by pekka2** [comment]
@ADDCreative 
Can this is better.:
` if(str_replace('\\', '/', realpath(DIR_IMAGE)) == str_replace('\\', '/', DIR_IMAGE) && !is_file(str_replace('\\', '/', realpath(DIR_IMAGE . $filename)) || !is_file(DIR_IMAGE . $filename)) {`
If realpath is not work, return old scipt

**Comment by ADDCreative** [comment]
@pekka2 Doing that would be pointless and would not prevent path traversal. 

Did you try the following?
```php
if (!is_file(DIR_IMAGE . $filename) || str_replace('\\', '/', realpath(DIR_IMAGE) . '/' . $filename) !== str_replace('\\'. '/', realpath(DIR_IMAGE . $filename))) {
```

**Comment by pekka2** [comment]
> Did you try the following?

No, I would not want to do tests very often my client server.

**Comment by pekka2** [comment]
Ps. But this 1 option is always true, if filename is not found. Next safety option never checked.

**Comment by ADDCreative** [comment]
@pekka2 If the file does not exist then there is no need to check the path as the method just ends without doing anything.

**Comment by pekka2** [comment]
@ADDCreative 
Issue this do, when image found 1. option: 
` !is_file(DIR_IMAGE . $filename)`
but is not found this:
`substr(str_replace('\\', '/', realpath(DIR_IMAGE . $filename)), 0, strlen(DIR_IMAGE)) != str_replace('\\', '/', DIR_IMAGE)`
